brave-http-tests/src/main/java/brave/http/ITHttpClient.java

```diff
@@ -0,0 +1,250 @@
+    try (SpanInScope ws = tracing.tracer().withSpanInScope(parent)) {
+      get(client, "/foo");
+    } finally {
+      parent.finish();
+    }
+
+    RecordedRequest request = server.takeRequest();
+    assertThat(request.getHeader("x-b3-traceId"))
+        .isEqualTo(parent.context().traceIdString());
+    assertThat(request.getHeader("x-b3-parentspanid"))
+        .isEqualTo(HexCodec.toLowerHex(parent.context().spanId()));
+  }
+
+  /**
+   * This tests that the parent is determined at the time the request was made, not when the request
+   * was executed.
+   */
+  @Test
+  public void usesParentFromInvocationTime() throws Exception {
+    server.enqueue(new MockResponse().setBodyDelay(1, TimeUnit.SECONDS));
+    server.enqueue(new MockResponse());
+
+    brave.Span parent = tracing.tracer().newTrace().name("test").start();
+    try (SpanInScope ws = tracing.tracer().withSpanInScope(parent)) {
+      getAsync(client, "/foo");
+      getAsync(client, "/foo");
+    } finally {
+      parent.finish();
+    }
+
+    brave.Span otherSpan = tracing.tracer().newTrace().name("test2").start();
+    try (SpanInScope ws = tracing.tracer().withSpanInScope(otherSpan)) {
+      for (int i = 0; i < 2; i++) {
+        RecordedRequest request = server.takeRequest();
+        assertThat(request.getHeader("x-b3-traceId"))
+            .isEqualTo(parent.context().traceIdString());
+        assertThat(request.getHeader("x-b3-parentspanid"))
+            .isEqualTo(HexCodec.toLowerHex(parent.context().spanId()));
+      }
+    } finally {
+      otherSpan.finish();
+    }
+  }
+
+  /** Unlike Brave 3, Brave 4 propagates trace ids even when unsampled */
+  @Test
+  public void propagates_sampledFalse() throws Exception {
+    tracing = tracingBuilder(Sampler.NEVER_SAMPLE).build();
+    close();
+    client = newClient(server.getPort(), new HttpClientParser<>());
+
+    server.enqueue(new MockResponse());
+    get(client, "/foo");
+
+    RecordedRequest request = server.takeRequest();
+    assertThat(request.getHeaders().toMultimap())
+        .containsKeys("x-b3-traceId", "x-b3-spanId")
+        .doesNotContainKey("x-b3-parentSpanId")
+        .containsEntry("x-b3-sampled", asList("0"));
+  }
+
+  @Test
+  public void reportsClientAnnotationsToZipkin() throws Exception {
+    server.enqueue(new MockResponse());
+    get(client, "/foo");
+
+    assertThat(collectedSpans())
+        .flatExtracting(s -> s.annotations)
+        .extracting(a -> a.value)
+        .containsExactly("cs", "cr");
+  }
+
+  @Test
+  public void defaultSpanNameIsMethodName() throws Exception {
+    server.enqueue(new MockResponse());
+    get(client, "/foo");
+
+    assertThat(collectedSpans())
+        .extracting(s -> s.name)
+        .containsExactly("get");
+  }
+
+  @Test
+  public void supportsPortableCustomization() throws Exception {
+    close();
+    client = newClient(server.getPort(), new HttpClientParser<Req, Resp>(){
```

Comment:
@llinder @devinsba @hyleung the tradeoffs come out here.. `HttpClientParser` is generically parameterized, but.. it can be used for any type. For example, the spanName below gives you the ability to adapt the raw type into a generic http model, but you can also access the raw request directly should you need to.

While the adapt (maybe we should call it convert or translate) is a bit explicit, it has the advantage of more concise and direct code in the instrumentation, and also a minor point which is that you don't need to allocate objects for every request and response (because this is converting an existing object vs wrapping it)